File: main/SoftwareSocket.js
import Renderer from './Renderer'
import Software from './Software'

export default class SoftwareSocket {
  constructor (io, server) {
    if (!!this.constructor.instance) {
      return this.constructor.instance
    }

    this.constructor.instance = this

    this._server = server
    this._softwares = {}

    this._io = io.of('/software').on('connection', (socket) => {
      console.log('----- software connected -----')
      console.log(`software.id: ${socket.id}`)

      const software = new Software(socket)
      // this._server.client = software
      this._softwares[socket.id] = software
      console.log(this._softwares)

      socket.on('software', data => {
        software.software = data.software
        software.scene = data.scene
        software.saved = data.saved
        this.sendSoftwares()
      })

      socket.on('disconnect', () => {
        console.log('----- software disconnected -----')
        console.log(`software.id: ${socket.id}`)
        delete this._softwares[socket.id]
        console.log(this._softwares)
        this.sendSoftwares()
      })
    })
  }

  get softwares () { return this._softwares }

  sendSoftwares () {
    const softs = {}
    for (const id in this._softwares) {
      const soft = this._softwares[id].formatForRender()
      softs[id] = soft
    }

    const render = new Renderer()
    render.sendMessageMainData('softwares', softs)
  }
}

File: python/software_socket.py
import os
import sys
import logging
import json

parent_dir = os.path.dirname(__file__)
if parent_dir not in sys.path:
    sys.path.append(parent_dir)

vendor_dir = os.path.join(parent_dir, 'vendor')
if vendor_dir not in sys.path:
    sys.path.append(vendor_dir)

import socketio

class Software(socketio.Namespace):
    def __init__(self, namespace, pulsar):
        super(Software, self).__init__(namespace)
        self._pulsar = pulsar

    def on_connect(self, sid, environ):
        print("----- connected software -----", sid)
        self._pulsar._softwares[sid] = {
            "id": sid,
            "software": None,
            "scene": None
        }

    def on_software(self, sid, data):
        print(sid, "----- software -----", data)
        data["id"] = sid
        self._pulsar._softwares[sid] = data

        if not self._pulsar._frontend == None:
            print('----- sendings software list to frontend -----')
            self._pulsar._sio.emit("softwares", self._pulsar._softwares, namespace="/frontend")

    def on_saved(self, sid, data):
        print("----- software saved -----", data)
        self._pulsar._softwares[sid]["saved"] = data
        if not self._pulsar._frontend == None:
            print('----- sendings software list to frontend -----')
            self._pulsar._sio.emit("softwares", self._pulsar._softwares, namespace="/frontend")

    def on_close(self, sid):
        self._pulsar._sio.disconnect(sid, namespace="/software")

    def on_disconnect(self, sid):
        print("----- disconnected -----", sid)
        del self._pulsar._softwares[sid]
        if not self._pulsar._frontend == None:
            print('----- sendings software list to frontend -----')
            self._pulsar._sio.emit("softwares", self._pulsar._softwares, namespace="/frontend")
